packages/core/src/fn.ts

```diff
@@ -1,22 +1,157 @@
+import * as ffi from "./ffi.js";
 import { Int } from "./int.js";
+import * as real from "./real.js";
+import { Real } from "./real.js";
+import { Val } from "./val.js";
 
-declare const opaqueBool: unique symbol;
-export type Bools = string & { tag: typeof opaqueBool };
-
-declare const opaqueInt: unique symbol;
-export type Ints = string & { tag: typeof opaqueInt };
-
-declare const opaqueReal: unique symbol;
-export type Reals = string & { tag: typeof opaqueReal };
+export type Bools = { tag: "Bool" };
+export type Ints = { tag: "Int" };
+export type Reals = { tag: "Real" };
 
 export interface Vecs<T> {
+  tag: "Vec";
   t: T;
   n: Int;
 }
 
 export type Type = Bools | Ints | Reals | Vecs<Type>;
 
-export interface Fn<T> {
-  params: [symbol, Type][];
-  body: T;
+export interface Fn {
+  params: Type[];
+  f: ffi.Fn;
+}
+
+type Resolve<T extends readonly Reals[]> = {
+  [K in keyof T]: Real;
+};
+
+class Body {
+  locals: Map<unknown, number>;
+  instrs: ffi.Body;
+
+  constructor(params: symbol[]) {
+    this.instrs = new ffi.Body();
+    try {
+      for (let i = params.length - 1; i >= 0; --i) this.instrs.set(i);
+      this.locals = new Map(params.map((param, id) => [param, id]));
+    } catch (e) {
+      this.instrs.free();
+      throw e;
+    }
+  }
+
+  local(x: unknown): number {
+    const id = this.locals.size;
+    this.locals.set(x, id);
+    this.instrs.set(id);
+    this.instrs.get(id);
+    return id;
+  }
+
```

Comment:
kind confused about what's going on here with `val<T>`, is this for branching cases?

Reply:
Yeah, `Val` is for kinds of nodes that can appear regardless of type, such as `symbol`s representing function parameters, or `Call`s to other functions, or `Index`ing into a vector, or `If`/else conditionals.